src/core/distributions/InverseDistribution.h

```diff
@@ -0,0 +1,154 @@
+#ifndef InverseDistribution_h
+#define InverseDistribution_h
+
+#include "TypedDistribution.h"
+#include "TypedDagNode.h"
+#include <memory>
+
+namespace RevBayesCore {
+    
+    /**
+     * Provides the inverse of the probability of a distribution.
+     *
+     * @copyright Copyright 2024-
+     * @author Martin R. Smith
+     * @since 2024-09-17, version 1.2.5
+     *
+     */
+    template<typename valType>
+    class InverseDistribution : public TypedDistribution< valType > {
+     
+    public:
+        // constructor(s)
+        InverseDistribution(const TypedDistribution<valType>& d)
+            : TypedDistribution<valType>( nullptr ),
+              dist( d.clone() )
+        {
+            // add the parameters to our set (in the base class)
+            // in that way other class can easily access the set of our parameters
+            // this will also ensure that the parameters are not getting deleted before we do
+            
+            // add the parameters of the distribution
+            for (const auto& parameter : dist->getParameters()) {
+                // Doesn't just add the parameter, but also links it so that updates cause
+                // the distribution to update
+                this->addParameter( parameter );
+            }
+            
+	        redrawValue();
+        }	
+
+        InverseDistribution(const InverseDistribution &d)
+            : TypedDistribution<valType>( d ),
+              dist( d.dist->clone() )
+        {
+            // add the parameters to our set (in the base class)
+            // in that way other class can easily access the set of our parameters
+            // this will also ensure that the parameters are not getting deleted before we do
+            
+            // add the parameters of the distribution
+            for (const auto& parameter : dist->getParameters()) {
+                this->addParameter( parameter );
+            }
+        }
+
+        // Functions from Distribution.cpp
+        void touch(const DagNode *affecter, bool touchAll) override
```

Comment:
Guessing a little bit here, because I'm not so familiar with how the touching process is handled, but I wonder whether this has something to do with a cloning happening somewhere?  e.g.  does `dnInverse(dnSomething(x))` create a clone (or perhaps copy) of `dnSomething()` that wouldn't be registered for update in the usual fashion?

I assume that direct adjustments to e.g. `dist->value` from within `InverseDistribution` don't need registering separately.

Reply:
If it gives a bit more context, the 'bad behaviour' case that this fixes is:

```rb
k ~ dnUniform(1, 10)
kRates := [ [ 0, k],
            [ 1, 0] ]
qK := fnFreeK(kRates)
invPhy ~ dnInv(dnPhyloCTMC(tree = phylogeny, Q = qK, type = "Standard"))
invPhy.clamp(chars)

k.clamp(1)
invPhy.lnProbability() # Calculated correctly

k.clamp(5) # This did not mark nodes as dirty until `touch` was added
invPhy.lnProbability() # Hence this was not calculated correctly
```

Reply:
@bjoelle Do you remember why the dnInverse would get touched when the contained `dists` was not touched?

I guess if you run this in the debugger and break on the `touch( )`, you would find out who is calling `touch( )`.

But it might be easier just to add a note saying that somehow dnInverse::dist wasn't getting touched, and this fixes it, but possible there is a better solution.  Or something like that.

Reply:
OK so one example where `dnInverse` is touched is:

```rb
inv_exp ~ dnInv(dnExp(1))
inv_exp.clamp(42) # Touches inv_exp twice
```

Call stack for first touch:
`StochasticNode[inv_exp]::clamp()` → `StochasticNode[inv_exp]::setValue` → `DagNode::touch()` → `StochasticNode::touchMe` → `InverseDistribution::touch()`

Call stack for second touch:
`evaluaateContent` → `MemberProcedure::execute()` → `DagNode::touch()` (and as before).  It's not clear that `dist.touch()` would have been triggered in this case.

Reply:
We see an equivalent pair of `touch`es with 
```rb
invPhy ~ dnInv(dnPhyloCTMC(tree = phylogeny, Q = qK, type = "Standard"))
invPhy.clamp(chars)
```
1. ![image](https://github.com/user-attachments/assets/81adf2a3-fc03-4e04-b14c-60ca77db1da9)
2. ![image](https://github.com/user-attachments/assets/4c955a3f-6cc2-46d9-8759-d8a2641f1430)

Reply:
Thanks for looking into this!  I don't think we need to solve everything now, but it may be good to save the partial progress as comments in the code for the next person who needs to look at this.  Maybe add a note that touch is called from 'clamp' as well as from invalidated children?

Reply:
Good suggestion; I've added a precis of the above into the comments.